File: src/exercicio-computacional-2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <stack>
#include <thread>
#include <chrono>
#include <mutex>
#include <atomic>

using namespace std;
mutex labirinto_mutex;  // Mutex para sincronizar o acesso ao labirinto
atomic<bool> encontrou_saida(false);  // Flag para indicar se a saída foi encontrada

// Função para carregar o labirinto a partir de um arquivo
bool carregar_labirinto(const string& caminho_arquivo, vector<vector<char>>& labirinto, int& linhas, int& colunas, pair<int, int>& inicio) {
    ifstream arquivo(caminho_arquivo);
    if (!arquivo.is_open()) {
        cerr << "Erro ao abrir o arquivo." << endl;
        return false;
    }

    arquivo >> linhas >> colunas;
    labirinto.resize(linhas);
    string linha;
    getline(arquivo, linha); // Para ler o '\n' após as dimensões

    for (int i = 0; i < linhas; i++) {
        getline(arquivo, linha);
        labirinto[i] = vector<char>(linha.begin(), linha.end());
        size_t pos = linha.find('e');
        if (pos != string::npos) {
            inicio = {i, static_cast<int>(pos)};
        }
    }

    arquivo.close();
    return true;
}

// Função para imprimir o labirinto
void imprimir_labirinto(const vector<vector<char>>& labirinto) {
    this_thread::sleep_for(chrono::milliseconds(100));

    // Limpa a tela no terminal
    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif

    for (const auto& linha : labirinto) {
        for (const auto& celula : linha) {
            cout << celula;
        }
        cout << endl;
    }
}

// Função recursiva para resolver o labirinto
bool resolver_labirinto(vector<vector<char>>& labirinto, int linhas, int colunas, pair<int, int> inicio) {
    stack<pair<int, int>> pilha;
    pilha.push(inicio);

    while (!pilha.empty()) {
        if (encontrou_saida.load()) {
            // Se a saída foi encontrada, esta thread termina
            return false;
        }

        auto [x, y] = pilha.top();
        pilha.pop();

        // Verifica se é a saída
        if (labirinto[x][y] == 's') {
            {
                lock_guard<mutex> guard(labirinto_mutex);
                if (!encontrou_saida.load()) {
                    encontrou_saida.store(true);  // Marca que a saída foi encontrada
                    labirinto[x][y] = 'o';
                    cout << "Saída encontrada!\n";
                    imprimir_labirinto(labirinto);
                }
            }
            return true;
        }

        if (labirinto[x][y] == 'x' || labirinto[x][y] == 'e') {
            lock_guard<mutex> guard(labirinto_mutex);
            labirinto[x][y] = 'o';
        }

        imprimir_labirinto(labirinto);

        {
            lock_guard<mutex> guard(labirinto_mutex);
            labirinto[x][y] = '.';
        }

        // Lista de possíveis direções (cima, baixo, esquerda, direita)
        vector<pair<int, int>> movimentos = {
            {x - 1, y}, {x + 1, y}, {x, y - 1}, {x, y + 1}
        };

        // Filtra movimentos válidos
        vector<pair<int, int>> caminhos_validos;
        for (const auto& [nx, ny] : movimentos) {
            if (nx >= 0 && nx < linhas && ny >= 0 && ny < colunas) {
                lock_guard<mutex> guard(labirinto_mutex);
                if (labirinto[nx][ny] != '#' && labirinto[nx][ny] != '.') {
                    caminhos_validos.push_back({nx, ny});
                }
            }
        }

        // Se houver mais de um caminho válido, explorar com threads
        if (caminhos_validos.size() > 1) {
            vector<thread> threads;

            // Cria threads para cada um dos caminhos válidos adicionais
            for (size_t i = 1; i < caminhos_validos.size(); i++) {
                threads.emplace_back([&labirinto, linhas, colunas, caminho = caminhos_validos[i]] {
                    resolver_labirinto(labirinto, linhas, colunas, caminho);
                });
            }

            // Explora o primeiro caminho na thread atual
            pilha.push(caminhos_validos[0]);

            // Espera todas as threads adicionais terminarem
            for (auto& t : threads) {
                if (t.joinable()) {
                    t.join();
                }
            }
        } else if (!caminhos_validos.empty()) {
            pilha.push(caminhos_validos[0]);
        }
    }

    if (!encontrou_saida.load()) {
        cout << "Nenhum caminho para a saída foi encontrado." << endl;
    }
    return false;
}


int main() {
    vector<vector<char>> labirinto;
    int linhas, colunas;
    pair<int, int> inicio;

    if (!carregar_labirinto("../data/maze2.txt", labirinto, linhas, colunas, inicio)) {
        return 1;
    }

    imprimir_labirinto(labirinto);
    cout << "Resolvendo labirinto...\n";

    resolver_labirinto(labirinto, linhas, colunas, inicio);

    return 0;
}

File: src/exercicio-computacional-2.js
"use strict";

const fs = require("fs");
const {
  Worker,
  isMainThread,
  parentPort,
  workerData,
} = require("worker_threads");

function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(ms);
    }, ms);
  });
}

function carregaArquivo(filename) {
  const conteudoLabirinto = fs.readFileSync(filename, "utf-8");
  const arrayLinhas = conteudoLabirinto.split("\n");

  const [tamanhoLinhas, tamanhoColunas] = arrayLinhas[0].split(" ").map(Number);

  const labirinto = [];
  let entradaLabirinto = null;

  // Encontrar entrada
  for (let i = 1; i <= tamanhoLinhas; i++) {
    const linha = arrayLinhas[i].split("");
    labirinto.push(linha);

    if (!entradaLabirinto) {
      const posicaoEntradaLabirinto = linha.indexOf("e");
      if (posicaoEntradaLabirinto !== -1) {
        entradaLabirinto = [i - 1, posicaoEntradaLabirinto];
      }
    }
  }

  return { labirinto, tamanhoLinhas, tamanhoColunas, entradaLabirinto };
}

async function printMatriz(labirinto) {
  await sleep(300);
  console.clear();
  labirinto.forEach((linha) => console.log(linha.join("")));
}

async function pesquisaSaida({
  labirinto,
  tamanhoLinhas,
  tamanhoColunas,
  entradaLabirinto,
}) {
  const posicaoAtual = [entradaLabirinto];

  while (posicaoAtual.length > 0) {
    const [x, y] = posicaoAtual.pop();

    if (labirinto[x][y] === "s") {
      console.log("Saída encontrada!");
      console.log("Posição da saida:", x, y, "\n");

      labirinto[x][y] = "o";
      await printMatriz(labirinto);

      return true;
    }

    // Marca a posição atual como visitada
    if (labirinto[x][y] === "x" || labirinto[x][y] === "e") {
      labirinto[x][y] = "o";
    }

    await printMatriz(labirinto);

    const direcoes = [
      [x - 1, y], // cima
      [x + 1, y], // baixo
      [x, y - 1], // esquerda
      [x, y + 1], // direita
    ];

    let caminhosPossiveis = [];
    for (const [nx, ny] of direcoes) {
      if (
        nx >= 0 &&
        ny >= 0 &&
        nx < tamanhoLinhas &&
        ny < tamanhoColunas &&
        (labirinto[nx][ny] === "x" || labirinto[nx][ny] === "s")
      ) {
        labirinto[x][y] = ".";
        caminhosPossiveis.push([nx, ny]);
      }
    }

    if (caminhosPossiveis.length > 1) {
      for (let i = 1; i < caminhosPossiveis.length; i++) {
        criarNovaThreadParaCaminho(
          labirinto,
          tamanhoLinhas,
          tamanhoColunas,
          caminhosPossiveis[i]
        );
      }
      posicaoAtual.push(caminhosPossiveis[0]);
    } else if (caminhosPossiveis.length === 1) {
      posicaoAtual.push(caminhosPossiveis[0]);
    }
  }

  console.log("Nenhum caminho para a saída foi encontrado.");
  return false;
}

function criarNovaThreadParaCaminho(
  labirinto,
  tamanhoLinhas,
  tamanhoColunas,
  inicio
) {
  const novoLabirinto = labirinto.map((linha) => [...linha]); // Faz uma cópia do labirinto
  const worker = new Worker(__filename, {
    workerData: {
      labirinto: novoLabirinto,
      tamanhoLinhas,
      tamanhoColunas,
      inicio,
    },
  });

  worker.on("message", (msg) => console.log(msg));
  worker.on("error", (err) => console.error(err));
  worker.on("exit", (code) => {
    if (code !== 0) {
      console.error(`Thread finalizada com código: ${code}`);
    }
  });
}

if (isMainThread) {
  // Função principal
  async function main() {
    const informacoesMatrizLabirinto = carregaArquivo("../data/maze2.txt");

    await printMatriz(informacoesMatrizLabirinto.labirinto);

    console.log("Resolvendo labirinto com multithreads...\n");

    await pesquisaSaida(informacoesMatrizLabirinto);
  }

  (async () => {
    await main();
  })();
} else {
  // Código executado nas threads
  const { labirinto, tamanhoLinhas, tamanhoColunas, inicio } = workerData;
  (async () => {
    await pesquisaSaida({
      labirinto,
      tamanhoLinhas,
      tamanhoColunas,
      entradaLabirinto: inicio,
    });
    parentPort.postMessage("Thread finalizada com sucesso.");
  })();
}
